skorch/net.py

```diff
@@ -336,19 +337,54 @@ def _yield_callbacks(self):
         print_logs = []
         for item in self.get_default_callbacks() + (self.callbacks or []):
             if isinstance(item, (tuple, list)):
+                named_by_user = True
                 name, cb = item
             else:
+                named_by_user = False
                 cb = item
                 if isinstance(cb, type):  # uninitialized:
                     name = cb.__name__
                 else:
                     name = cb.__class__.__name__
             if isinstance(cb, PrintLog) or (cb == PrintLog):
-                print_logs.append((name, cb))
+                print_logs.append((name, cb, named_by_user))
             else:
-                yield name, cb
+                yield name, cb, named_by_user
         yield from print_logs
 
+    def _callbacks_grouped_by_name(self):
+        """Group callbacks by name and collect names set by the user."""
+        callbacks, names_set_by_user = {}, set()
+        for name, cb, named_by_user in self._yield_callbacks():
+            if named_by_user:
+                names_set_by_user.add(name)
+            callbacks[name] = callbacks.get(name, []) + [cb]
+        return callbacks, names_set_by_user
+
+    def _uniquely_named_callbacks(self):
+        """Make sure that the returned dict of named callbacks is unique
+        w.r.t. to the callback name. User-defined names will not be
+        renamed on conflict, instead an exception will be raised. The
+        same goes for the event where renaming leads to a conflict.
+        """
+        grouped_cbs, names_set_by_user = self._callbacks_grouped_by_name()
+        for name, cbs in grouped_cbs.items():
+            if len(cbs) > 1 and name in names_set_by_user:
+                raise ValueError("Found duplicate user-set callback name "
+                                 "'{}'. Use unique names to correct this."
+                                 .format(name))
+
+            for i, cb in enumerate(cbs):
+                if len(cbs) > 1:
+                    unique_name = '{}_{}'.format(name, i+1)
+                    if unique_name in grouped_cbs:
+                        raise ValueError("Setting unique name failed since "
```

Comment:
As in the error message above, I would mention that we deal with *callback names* to make it extra clear.